packages/lesswrong/server/languageModels/languageModelIntegration.ts

```diff
@@ -0,0 +1,307 @@
+    return str;
+  
+  // First try an encode-then-decode roundtrip
+  try {
+    const encoded = gpt3encode(str);
+    
+    if (encoded.length <= tokens) return str;
+    const redecoded = gpt3decode(take(encoded,tokens));
+    if (redecoded === str.substring(0,redecoded.length)) {
+      return redecoded;
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`GPT-3 encoding did not roundtrip: ${JSON.stringify(str)}`);
+    }
+  } catch {
+    console.log(`Could not encode string for truncation length estimate: ${JSON.stringify(str)}`); //eslint-disable-line no-console
+  }
+  
+  // If that didn't work, binary-search string truncations to find one that has the right token count
+  let low=0, high=str.length, mid=(low+high)/2|0;
+  
+  while (high>low) {
+    if (countGptTokens(str.substring(0,mid)) > tokens) {
+      high = mid;
+    } else {
+      low = mid+1;
+    }
+    mid=(low+high)/2|0;
+  }
+  
+  return str.substring(0,mid);
+}
+
+/**
+ * Perform a binary-classification task using a language model. The provided
+ * input is the text that is being classified; it does *not* include the wrapping
+ * template, which is admin-configurable and loaded from the database.
+ */
+export async function languageModelClassify({taskName, inputs, context}: {
+  taskName: LanguageModelClassificationTask,
+  inputs: Record<string,string>,
+  context: ResolverContext,
+}): Promise<boolean|"maybe"> {
+  const lmConfig = await getLMConfigForTask(taskName, context);
+  const continuation = await languageModelExecute({
+    template: lmConfig.template,
+    inputs,
+    task: lmConfig.task,
+    api: lmConfig.api,
+    model: lmConfig.model,
+    maxTokens: 1,
+  });
+  const tokenizedContinuation = continuation.split(/\s+/);
+  const firstToken = tokenizedContinuation[0];
+  
+  if (firstToken.toLowerCase()==="yes")
+    return true;
+  else if (firstToken.toLowerCase()==="no")
+    return false;
+  else
+    return "maybe";
+}
+
+/**
+ * Perform a text-generation task using a language model. The provided input is
+ the text that is being classified; it does *not* include the wrapping template,
+ * which is admin-configurable and loaded from the database.
+ */
+export async function languageModelGenerateText({taskName, inputs, maxTokens, context}: {
+  taskName: LanguageModelGenerationTask,
+  inputs: Record<string,string>,
+  maxTokens: number,
+  context: ResolverContext,
+}): Promise<string> {
+  const lmConfig = await getLMConfigForTask(taskName, context);
+  const continuation = await languageModelExecute({
+    template: lmConfig.template,
+    inputs,
+    task: lmConfig.task,
+    api: lmConfig.api,
+    model: lmConfig.model,
+    maxTokens,
+  });
+  return continuation;
+}
+
+async function languageModelExecute(job: LanguageModelJob): Promise<string> {
+  // TODO: Check and populate the cache
+  
+  switch(job.api) {
+    default: {
+      throw new Error(`Invalid language model API: ${job.api}`);
+    }
```

Comment:
Isn't `default` usually last?  Also I think it might be redundant since the `api` field is constrained to `type LanguageModelAPI = "disabled"|"stub"|"openai";`

Reply:
Sure, moved to last (though I don't think it matters and our style doc doesn't say anything about it). Whether this is validation or defensive programming depends on how this got called; I believe this came from a user provided template (`lm-config-summarize`) and might or might not be the first actual check that the string is valid.

Reply:
Yeah I think I saw https://github.com/ForumMagnum/ForumMagnum/pull/6346/#discussion_r1057335398 after this